src/app/codegen-data-model/CodegenDataModel_Read.cpp

```diff
@@ -0,0 +1,370 @@
+        if (type == nullptr)
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedEndpoint);
+        }
+
+        const EmberAfCluster * cluster = emberAfFindClusterInType(type, aPath.mClusterId, CLUSTER_MASK_SERVER);
+        if (cluster == nullptr)
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedCluster);
+        }
+
+        // Since we know the attribute is unsupported and the endpoint/cluster are
+        // OK, this is the only option left.
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+    }
+
+    return metadata;
+}
+
+/// Attempts to read via an attribute access interface (AAI)
+///
+/// If it returns a CHIP_ERROR, then this is a FINAL result (i.e. either failure or success).
+///
+/// If it returns std::nullopt, then there is no AAI to handle the given path
+/// and processing should figure out the value otherwise (generally from other ember data)
+std::optional<CHIP_ERROR> TryReadViaAccessInterface(const ConcreteAttributePath & path, AttributeAccessInterface * aai,
+                                                    AttributeValueEncoder & encoder)
+{
+    // Processing can happen only if an attribute access interface actually exists..
+    if (aai == nullptr)
+    {
+        return std::nullopt;
+    }
+
+    CHIP_ERROR err = aai->Read(path, encoder);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        // Implementation of 8.4.3.2 of the spec for path expansion
+        if (path.mExpanded && (err == CHIP_IM_GLOBAL_STATUS(UnsupportedRead)))
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        return err;
+    }
+
+    // If the encoder tried to encode, then a value should have been written.
+    //   - if encode, assume DONE (i.e. FINAL CHIP_NO_ERROR)
+    //   - if no encode, say that processing must continue
+    return encoder.TriedEncode() ? std::make_optional(CHIP_NO_ERROR) : std::nullopt;
+}
+
+/// Metadata of what a ember/pascal short string means (prepended by a u8 length)
+struct ShortPascalString
+{
+    using LengthType                        = uint8_t;
+    static constexpr LengthType kNullLength = 0xFF;
+};
+
+/// Metadata of what a ember/pascal LONG string means (prepended by a u16 length)
+struct LongPascalString
+{
+    using LengthType                        = uint16_t;
+    static constexpr LengthType kNullLength = 0xFFFF;
+};
+
+// ember assumptions ... should just work
+static_assert(sizeof(ShortPascalString::LengthType) == 1);
+static_assert(sizeof(LongPascalString::LengthType) == 2);
+
+/// Given a ByteSpan containing data from ember, interpret it
+/// as a span of type OUT (i.e. ByteSpan or CharSpan) given a ENCODING
+/// where ENCODING is Short or Long pascal strings.
+template <class OUT, class ENCODING>
+std::optional<OUT> ExtractEmberString(ByteSpan data)
+{
+    typename ENCODING::LengthType len;
+
+    VerifyOrDie(sizeof(len) <= data.size());
+    memcpy(&len, data.data(), sizeof(len));
```

Comment:
This does not account for endianness, but IIRC the endianness is native for the 16 bit lengths. Suggest commenting on data format assumptjons for Pascal strings, especially since direct native bytes are copied into scalars.

Reply:
Will add comments. this is indeed "ember internal format" which uses native byte order.